lib/logstash/inputs/decoder_impl.rb

```diff
@@ -0,0 +1,63 @@
+# encoding: utf-8
+require 'java'
+
+java_import org.logstash.tcp.InputLoop
+java_import org.logstash.tcp.Decoder
+
+class DecoderImpl
+
+  include Decoder
+
+  def initialize(codec, tcp)
+    @tcp = tcp
+    @codec = codec
+    @codecs = {}
+    @first_read = {}
+    @addresses = {}
+    @ports = {}
+    @proxy_addresses = {}
+    @proxy_ports = {}
+  end
+
+  def decode(channel_addr, data)
+    if @codecs[channel_addr].nil?
+      @codecs[channel_addr] = @codec.clone
+      @first_read[channel_addr] = true
+    end
+    bytes = Java::byte[data.readableBytes].new
+    data.getBytes(0, bytes)
+    tbuf = String.from_java_bytes bytes, "ASCII-8BIT"
```

Comment:
@jakelandis `ASCII-8BIT` is raw bytes basically. If it's some other encoding, then that will be enforced down the line by the `@codec`. => `ASCII-8BIT` always works and will never coerce anything about the encoding  :) (Ruby is weird :D I know)